protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -81,148 +51,240 @@ where
 {
     type ConnectionHandler = Handler;
 
-    type ToSwarm = Event;
+    type ToSwarm = ();
 
     fn handle_established_inbound_connection(
         &mut self,
-        _connection_id: ConnectionId,
-        peer: PeerId,
+        connection_id: ConnectionId,
+        _peer: PeerId,
         _local_addr: &Multiaddr,
-        _remote_addr: &Multiaddr,
-    ) -> Result<Self::ConnectionHandler, ConnectionDenied> {
-        Ok(Handler::new(self.accepted_nonce.clone(), peer))
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if addr_is_local(remote_addr) {
+            self.local_peers.insert(connection_id);
+        }
+        Ok(new_handler())
     }
 
     fn handle_established_outbound_connection(
         &mut self,
-        _connection_id: ConnectionId,
-        peer: PeerId,
-        _addr: &Multiaddr,
+        connection_id: ConnectionId,
+        _peer: PeerId,
+        addr: &Multiaddr,
         _role_override: Endpoint,
         _port_use: PortUse,
-    ) -> Result<Self::ConnectionHandler, ConnectionDenied> {
-        Ok(Handler::new(self.accepted_nonce.clone(), peer))
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if addr_is_local(addr) {
+            self.local_peers.insert(connection_id);
+        }
+        Ok(new_handler())
     }
 
-    fn on_swarm_event(&mut self, _event: FromSwarm) {
-        todo!()
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::NewExternalAddrCandidate(NewExternalAddrCandidate { addr }) => {
+                for _ in 0..self.config.test_server_count {
+                    let nonce = self.rng.gen();
+                    self.pending_requests.push_back(DialRequest {
```

Comment:
I'd suggest to directly construct the `ToSwarm` event here instead of an indirection via a pending request.

Reply:
(Only if we are connected though)

Reply:
But that is kind of obsolete with the throttling proposed by you previously.

Reply:
This particular instance maybe, yes. It was more meant as a general suggestion :)